DCMAW-11008: Add test harness documentation (#408)
govuk-one-login/mobile-id-check-async: test-resources/src/functions/dequeue/README.md

Reviewer: suggestion: DEQUEUE.md?

Reviewer: suggestion: move this into the docs folder - I don't know if this should live in the `src` directory

Author: Renamed and moved to docs folder

# Dequeue events

## Contents
- [Overview](#overview)
- [Dependencies](#dependencies)
- [Dequeue events flow](#dequeue-events-flow)
- [How it works](#how-it-works)
  - [Backend API SQS queue](#backend-api-sqs-queue)
  - [Dequeue Lambda](#dequeue-lambda)
    - [Lambda invocation](#lambda-invocation)
    - [Processing events](#processing-events)
    - [Storing events in DynamoDB](#storing-events-in-dynamodb)
    - [`BatchItemFailures`](#batchitemfailures)
- [Troubleshooting](#troubleshooting)

## Overview

#### What problem does this solve?

> TxMA is a queue based system where the Async Backend application sends audit

Reviewer: suggestion: Please may you give this another go at crafting this?

I think the first sentence will be easier to write if it starts: "The async backend sends"

Author: updated

events to SQS, which then triggers a Lambda within an AWS account. Though the

> flow of an event can be tested manually, it is not possible to do the same with

Reviewer: nit:

Please may we be specific about what we mean by 'flow'? One _could_ argue that unit testing the lambda sending the event to TxMA tests the flow.

Author: updated

unit testing. The Dequeue events functionality provides an automated solution.

> #### Why do we need to test this?

Reviewer: suggestion: Fold this into the title above.

I think the problem and why we need to test this are the same thing: just worded differently. 

This section (folded) with the one above could read as follows:

-> Context of the solution (as you've done)
-> The problem
-> The solution

Author: updated

The purpose of the Dequeue flow is to switch from manual testing to an automated
testing strategy, providing a way to assert that TxMA audit events sent to the
backend-api SQS queue are 1) successfully added to the queue, and 2) are in the
correct shape.

It achieves this by writing events to a database, so that they can later be
retrieved and tested against, adding the ability to test two existing
patterns within the Async repo architecture:

- TxMA audit events sent to SQS
-  Messages sent to the following during handback:
    - Vendor processing queue

>     - Outbound queue for IPV

Author: I've removed this due to repetition.

#### How does it work?

To enable assertions on audit events, this test harness uses an AWS resource to
map a Lambda function to an SQS queue. An event triggers the Lambda to pull one
or more *`Records` from the queue to process and store in an Events table in
DynamoDB.

> > *Note: a `Record` is the data structure that contains a payload, e.g., an

Reviewer: nit: I think Record is potentially a bit specific here and probably not for the summary

Author: Updated

> **audit event**.

Reviewer: nit: 'to enable assertions on audit events' doesn't kick this paragraph off as clearly as we could. We need to explain the _how_. 

I would give a really condensed overview of the flow. 

e.g. A backend lambda writes a message to Txma SQS. The dequeue lambda then asynchronously polls the queue and reads messages in batches. The lambda saves the event payload in Dynamo. The events api provides the ability to query Dynamo to retrieve one or more events. This events api can be used to...

Author: updated

## How it works

<img src="events_flow_diagram.png">

> ###### Dequeue events flow

##

### Backend API SQS queue

1. #### Writing to TxMA SQS

> The first step in the Dequeue events flow happens when a TxMA audit event is

Reviewer: nit: we've got 1. in the title - we've already said it's the first step.

I don't think we need to specify whether it's done manually or in a deployed stack.

Author: Updated

sent to the backend-api SQS queue. This can be done manually in the AWS console
and by calling any Lambda, in a deployed backend-api stack, that pushes a TxMA
event to SQS.

2. #### Async Backend API TxMA SQS queue
Once events reach SQS, they are then pulled off of the queue by the Dequeue
Lambda to be processed.

##

### [Dequeue Lambda](./dequeueHandler.ts)

3. #### Lambda invocation

The Dequeue Lambda receives an event when a new event is added to the
backend-api SQS queue. This is done using an [`EventSourceMapping` AWS resource](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-eventsourcemapping.html).

> 

Reviewer: suggestion: Lambda polls SQS for messages

Author: Updated

> 4. #### Processing events

Reviewer: suggestion: I wonder whether this is better phrased as Event validation?

Author: updated

> A new event is sent from the backend-api SQS queue to the Dequeue Lambda.

Reviewer: nit: We've covered this in step 1

Author: updated

Multiple TxMA events can be sent, in a `Records` array, by configuring the
`BatchSize` on the `EventSourceMapping` resource (an example can be found in the

> [SAM template](../../../infra/dequeue/function.yaml)).

Reviewer: nit: I think this is more closely aligned to previous sections, rather than about how the lambda processes events, once the invocation has started

Each Record is processed individually with an initial check that logs if there

> is an error retrieving the event from the Record. The Lambda will then skip to

Reviewer: question: what does 'an error retrieving the event from the Record' mean?

Is this simply: The lambda processes each event in the batch and validates whether it is an allowed event (potentially it does other things)?

If there is an error, the lambda...

Author: Updated

the next event to be processed if there is one.

5. #### Storing events in DynamoDB

> An event that has passed the previous check is then sent to the Events

Reviewer: ```suggestion
The lambda saves the event to the Events table via...
```

Author: Updated

table via a DynamoDB `PutItemCommand`. This call contains the following `Item`
schema:

```typescript
const putItemCommandInput: PutItemCommandInput = {
  TableName: env.EVENTS_TABLE_NAME,
  Item: marshall({
    pk: `SESSION#${sessionId}`,
    sk: `TXMA#EVENT_NAME#${eventName}#TIMESTAMP#${timestamp}`,
    event: record.body,
    timeToLiveInSeconds,
  }),
};
```

> ###### Partition Key and Sort Key
> The Partition Key (PK) and Sort Key (SK) make up the composite key use to
> query the Events table.

An error writing to DynamoDB results in a message being logged and the
`messageId` from the current record being pushed to a `batchItemFailures` array.

If storing the event in the Events table is successful, the `event_name` and
`session_id` from that event is pushed to a `processedMessages` array, which is
logged once all events have been processed.

> #### `BatchItemFailures`

Reviewer: question: Is this part of 5. Storing events in Dynamo?

Author: I've moved this to a new step

> An object with `batchItemFailures`, if any, is returned from the Dequeue Lambda.

Reviewer: nit: I think we always return this array?

Author: Updated
